### PR TITLE
[train] New persistence mode: Support the `tune.run(restore="ckpt_path")` API

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

`tune.run(restore="/path/to/local/checkpoint")` previously didn't work. This PR allows you to resume training a single trial from a checkpoint. However, this does NOT load back the training iteration from before. It just starts from scratch with the checkpoint. It's more similar to `Trainer(resume_from_checkpoint)` now. This is a behavior change due to the "trainable metadata" (including the training iter) no longer being saved within the checkpoint.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]

**Comment by justinvyu** [comment]
@krfricke ok to merge if looks good. Are you ok with the slight API behavior change here?

**Comment by ericl** [comment]
ping @zhe-thoughts for approval for 2.7

File: python/ray/tune/experiment/trial.py
Comment (krfricke) [inline python/ray/tune/experiment/trial.py:1097]: Is this missing a return?

```suggestion
    return

```

**Review by krfricke — APPROVED** [review]
LGTM, will merge after approval from @zhe-thoughts

**Review by zhe-thoughts — APPROVED** [review]
This is part of Train API clean up

**Comment by zhe-thoughts** [comment]
@krfricke or @ericl : is CI OK?

**Comment by zhe-thoughts** [comment]
Checked with @aslonnie and confirmed CI is OK

**Comment by zhe-thoughts** [comment]
Checked with @aslonnie and confirmed CI is OK

**Pull request merged**
